scalardl/src/jepsen/scalardl.clj

```diff
@@ -0,0 +1,159 @@
+(ns jepsen.scalardl
+  (:require [clojure.tools.logging :refer [debug info warn]]
+            [jepsen
+             [cli :as cli]
+             [client :as client]
+             [checker :as checker]
+             [control :as c]
+             [db :as db]
+             [generator :as gen]
+             [independent :as independent]
+             [nemesis :as nemesis]
+             [tests :as tests]]
+            [jepsen.control.util :as cu]
+            [jepsen.scalardl.support :as s]
+            [jepsen.os.debian :as debian]
+            [knossos.model :as model]
+            [knossos.op])
+  (:import (com.scalar.client.service ClientService)
+           (java.util Optional)))
+
+(defn parse-long
+  [s]
+  (when s (Long/parseLong s)))
+
+(def dir "/opt/scalardl")
+(def binary "bin/scalar-ledger")
+(def logfile (str dir "/scalardl.log"))
+(def pidfile (str dir "/scalardl.pid"))
+(def ledger-dir "/jepsen/scalardl/ledger")
+
+(defn r [_ _] {:type :invoke, :f :read, :value nil})
+(defn w [_ _] {:type :invoke, :f :write, :value (rand-int 5)})
+
+(defrecord Client [conn]
+  client/Client
+  (setup! [this test]
```

Comment:
These registrations work on each node.
Only the first request succeeds and others fail because the same certificate or contracts can't be registered by the same user?
I think you can limit to only the 1 client to register these registrations.

Reply:
Yes, this part also bothered me, but in the `setup!` phase I cannot specify a node. (Well, I can but it says it is deprecated.) In the `open!` phase Jepsen says that "Open *should not* affect the logical state of the test; it should not, for instance, modify tables or insert records." Is registering contracts and certifications affecting the logical state of a test? One last place to try is in the db phase after the first node has started. I think I tried this, but I'll give this a look again.

Reply:
Right, so the registering of certificates has to be done in the `Client`, there is no connection otherwise. So that leaves `setup` and `open`. Open is done before setup and is meant to open the connection. Setup does any required setup. This is done for every client, which is why when running 10 concurrent threads, we try to register certificates and contracts 10 times. Before trying to register certificates, etc., I could do a `listContract` and see if there is any data and then abort if so, but trying to register again doesn't create any problem. What do you think?

Reply:
I added an atom to the test so that only client will register the certificate contracts. It seems to work okay, but, as a solution, it doesn't feel very good.